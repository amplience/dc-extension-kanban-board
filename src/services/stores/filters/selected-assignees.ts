import { persistedWritable } from '../persisted-store';

export const selectedAssignees = persistedWritable<string[]>(
  'selectedAssignees',
  []
);
