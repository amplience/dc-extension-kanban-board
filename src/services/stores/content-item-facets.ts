import type { User } from 'dc-extensions-sdk/dist/types/lib/components/Users';
import type { FacetField } from 'dc-management-sdk-js';
import { derived } from 'svelte/store';
import { selectedAssignees } from './filters/selected-assignees';

export const contentItemFacets = derived(
  [selectedAssignees],
  ([$selectedAssignees]) => {
    const facets: FacetField[] = [];
    if ($selectedAssignees.length > 0) {
      facets.push(createAssigneeFacet($selectedAssignees));
    }
    return facets;
  }
);

function createAssigneeFacet(users: User[]): FacetField {
  return {
    facetAs: 'ENUM',
    field: 'assignees',
    filter: { type: 'IN', values: users.map((user) => user.id) },
    name: 'assignees',
  };
}
