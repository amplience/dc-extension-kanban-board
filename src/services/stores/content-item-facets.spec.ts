import { contentItemFacets, createAssigneeFacet } from './content-item-facets';
import { selectedAssignees } from './filters/selected-assignees';

const users = ['USER_1', 'USER_2', 'USER_3'];

describe('content-item-facets - store value', () => {
  it('should build an array of facets', () => {
    selectedAssignees.set(users);

    contentItemFacets.subscribe((facets) => {
      expect(facets).toMatchSnapshot();
    });
  });

  it('should build an array of facets without assignees', () => {
    selectedAssignees.set([]);

    contentItemFacets.subscribe((facets) => {
      expect(facets).toMatchInlineSnapshot(`Array []`);
    });
  });
});

describe('content-item-facets - createAssigneeFacet', () => {
  it('should return an assignee facet', () => {
    expect(createAssigneeFacet(users)).toMatchInlineSnapshot(`
      Object {
        "facetAs": "ENUM",
        "field": "assignees",
        "filter": Object {
          "type": "IN",
          "values": Array [
            "USER_1",
            "USER_2",
            "USER_3",
          ],
        },
        "name": "assignees",
      }
    `);
  });
});
