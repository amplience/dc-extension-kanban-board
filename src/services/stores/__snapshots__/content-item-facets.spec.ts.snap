// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`content-item-facets - store value should build an array of facets 1`] = `
Array [
  Object {
    "facetAs": "ENUM",
    "field": "assignees",
    "filter": Object {
      "type": "IN",
      "values": Array [
        "USER_1",
        "USER_2",
        "USER_3",
      ],
    },
    "name": "assignees",
  },
]
`;

exports[`content-item-facets - store value should build an array of facets without assignees 1`] = `Array []`;
